lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+	public $scanner;
+
+
+	/**
+	 * Set of modifications to perform on input HTML document, built
+	 * via calls to the fluid API exposed by WP_HTML_Processor.
+	 *
+	 * @var WP_Tag_Modifications
+	 */
+	public $modifications;
+
+
+	public function __construct( $input ) {
+		$this->scanner = new WP_HTML_Scanner( $input );
+		$this->modifications = new WP_Tag_Modifications();
+	}
+
+
+	public function find( $query ) {
+		$this->commit_class_changes();
+		$this->descriptor = WP_Tag_Find_Descriptor::parse( $query );
+		if ( false === $this->scanner->scan( $this->descriptor ) ) {
+			return false;
+		}
+
+		return $this;
+	}
+
+
+	public function add_attribute( $name, $value ) {
+		$tag_name = $this->scanner->tag_name;
+		$insert_at = $tag_name->start + $tag_name->length;
+
+		$this->modifications->replacements[] = array(
+			$insert_at,
+			$insert_at,
+			' ' . self::serialize_attribute( $name, $value )
+		);
+
+		return $this;
+	}
+
+
+	public function add_class( $class_name ) {
+		$this->modifications->class_changes[ $class_name ] = true;
+	}
+
+
+	public function set_attribute( $name, $value ) {
+		if ( ! array_key_exists( WP_Tag_Find_Descriptor::comparable( $name ), $this->scanner->attributes ) ) {
+			return $this->add_attribute( $name, $value );
+		}
+
+		/** @var WP_HTML_Attribute_Token $existing */
+		$existing = $this->scanner->attributes[ $name ];
+
+		$this->modifications->replacements[] = array(
+			$existing->attribute_start,
+			$existing->attribute_end,
+			self::serialize_attribute( $name, $value )
+		);
+	}
+
+
+	public function remove_class( $class_name ) {
+		$this->modifications->class_changes[ $class_name ] = false;
+	}
+
+
+	public function remove_attribute( $name ) {
+		if ( ! array_key_exists( $name, $this->scanner->attributes ) ) {
+			return $this;
+		}
+
+		/** @var WP_HTML_Attribute_Token $existing */
+		$existing = $this->scanner->attributes[ $name ];
+
+		$this->modifications->replacements[] = array(
+			$existing->attribute_start,
+			$existing->attribute_end,
+			''
+		);
+	}
+
+
+	public static function serialize_attribute( $name, $value ) {
+		$value = str_replace( '"', "&quot;", $value );
+
+		return "{$name}=\"{$value}\"";
+	}
+
+
+	public function commit_class_changes() {
+		if ( empty( $this->modifications->class_changes ) ) {
+			return;
+		}
+
+		$existing_class = array_key_exists( 'class', $this->scanner->attributes )
+			? substr(
+				$this->scanner->document,
+				$this->scanner->attributes['class']->value->start,
+				$this->scanner->attributes['class']->value->length
+			)
+			: '';
+
+		// Remove unwanted classes.
+		$new_class = preg_replace_callback(
+			'~(?:^|[ \t])([^ \t]+)~miu',
+			function ( $matches ) {
+				list( $full_match, $class_name ) = $matches;
+
+				$comparable_name = WP_Tag_Find_Descriptor::comparable( $class_name );
+				if (
+					 array_key_exists( $comparable_name, $this->modifications->class_changes ) &&
+					 false === $this->modifications->class_changes[ $comparable_name ]
```

Comment:
```suggestion
					 array_key_exists( $comparable_name, $this->modifications->class_changes )
```

We can remove all the classes mentioned in `class_changes`. This way, the code that adds new classes a few lines later won't introduce duplicate names. Duplicates are fine, but why not avoid them if we can?

Reply:
you raise an interesting point. that's very true and a good idea to do, but also, I'd like to avoid changing any HTML we don't have to. that means this code as written does neither.

but, we could remember the names we scan when removing and then only add in the ones which weren't there.

```php
$existing_classes = array();
$without_removed_classes = preg_replace_callback(
    …,
    function ( $matches ) use ( &$existing_classes ) {
        …

        $existing_classes[ $comparable_name ] = true;

        …
    }
);

foreach ( $this->modifications->class_changes as $class => $should_include ) {
	if ( $should_include && ! array_key_exists( $existing_classes, $class ) ) {
		$new_class .= " {$class}";
	}
}
```

I think either case (removing all classes preemptively or only adding the missing ones) would work. From a consumer point of view I like the idea of leaving existing classes where they are if all we care about is adding them.

Reply:
I like leaving the existing classes where they are – let's remember the names we scan then!

Ah – and it would also be great to remove the `class` attribute entirely if there are no classes left.

Reply:
> Ah – and it would also be great to remove the class attribute entirely if there are no classes left.
an even better idea!